parser_library/src/lsp/folding.cpp

```diff
@@ -0,0 +1,282 @@
+        });
+
+        lineno = end_lineno;
+    }
+
+    return lines;
+}
+
+void mark_suspicious(std::vector<lsp::line_entry>& lines)
+{
+    static constexpr size_t min_lines = 50;
+    static constexpr size_t percentile_factor = 20; // 5%
+
+    std::vector<signed char> indents;
+    indents.reserve(lines.size());
+
+    for (const auto& e : lines)
+        if (e.indent > 0)
+            indents.push_back(e.indent);
+
+    if (indents.size() < min_lines)
+        return;
+
+    const auto percentile_pos = indents.begin() + indents.size() / percentile_factor;
+    std::nth_element(indents.begin(), percentile_pos, indents.end());
+    const auto limit = *percentile_pos;
+
+    for (auto& x : lines)
+        if (x.indent > 0 && x.indent < limit)
+            x.suspicious = true;
+}
+
+void folding_by_indentation(std::vector<fold_data>& data, std::span<const line_entry> lines)
+{
+    struct prev_region
+    {
+        size_t end_above;
+        size_t lineno;
+        signed char indent;
+    };
+
+    std::stack<prev_region> pr;
+    pr.push({ lines.back().end_lineno, lines.back().end_lineno, -1 });
+
+    for (auto it = lines.rbegin(); it != lines.rend(); ++it)
+    {
+        const auto& line = *it;
+
+        if (line.indent < 0 || line.comment || line.suspicious)
+            continue;
+
+        if (pr.top().indent > line.indent)
+        {
+            while (pr.top().indent > line.indent)
+                pr.pop();
+            if (pr.top().end_above >= 2 + line.end_lineno)
+                data[line.lineno].indentation = pr.top().end_above - 1;
+            else
+            {
+                data[line.lineno].small_structure = true;
+                data[pr.top().end_above - 1].small_structure = true;
+            }
+        }
+
+        if (pr.top().indent == line.indent)
+            pr.top().end_above = line.lineno;
+        else
+            pr.push({ line.lineno, line.lineno, line.indent });
+    }
+}
+
+void folding_by_comments(std::vector<fold_data>& data, std::span<const line_entry> lines)
+{
+    const auto iscomment = [](const line_entry& le) { return le.comment; };
+    for (auto it = lines.begin(); (it = std::find_if(it, lines.end(), iscomment)) != lines.end();)
+    {
+        auto end = std::find_if(it, lines.end(), std::not_fn(iscomment));
+
+        if (auto last = std::prev(end); it != last)
+            data[it->lineno].comment = last->end_lineno - 1;
+
+        it = end;
+    }
+}
+
+void folding_between_comments(std::vector<fold_data>& data, std::span<const line_entry> lines)
+{
+    const auto isnotcomment = [](const line_entry& le) { return !le.comment; };
+    for (auto it = lines.begin(); (it = std::find_if(it, lines.end(), isnotcomment)) != lines.end();)
+    {
+        auto end = std::find_if(it, lines.end(), std::not_fn(isnotcomment));
+
+        if (auto last = std::prev(end); it != last)
+            data[it->lineno].notcomment = last->end_lineno - 1;
+
+        it = end;
+    }
+}
```

Comment:
Consider shifting `iscomment()` to anonymous namespace and replacing `isnotcomment()` with `iscomment()` - one of the benefits would be not using "not" "not" syntax.

Also, I think it could be made `static consexpr`?

I guess, the logic can be extracted as well since it's very similar.

Something like:
```suggestion

namespace {
    static constexpr auto iscomment = [](const line_entry& le) { return le.comment; };  
    
    void comment_folding_helper(std::vector<fold_data>& data, std::span<const line_entry> lines, auto initCond, auto endCond) {
        for (auto it = lines.begin(); (it = std::find_if(it, lines.end(), initCond)) != lines.end();)
        {
            auto end = std::find_if(it, lines.end(), endCond);
    
            if (auto last = std::prev(end); it != last)
                data[it->lineno].comment = last->end_lineno - 1;
    
            it = end;
        }
    }
}

void folding_by_comments(std::vector<fold_data>& data, std::span<const line_entry> lines)
{
    comment_folding_helper(data, lines, iscomment, std::not_fn(iscomment));
}

void folding_between_comments(std::vector<fold_data>& data, std::span<const line_entry> lines)
{
    comment_folding_helper(data, lines, std::not_fn(iscomment), iscomment);
}
```

Reply:
Those functions each populate a different member `fold_data::comment` vs `fold_data::notcomment`.
As for the use of `not_fn`, I actually think it makes it very clear, that the inner `find_if` cannot return `it`.